### ISSUE TITLE
Chartmuseum apparently not closing files when scanning repo

### ISSUE THREAD

**Issue by sjthespian** [kubernetes-helm/chartmuseum#58]
It looks as if chartmuseum may not be closing files when it scans the repo. We are looking at converting a large repo (nearly 16000 packages) to chartmuseum and are getting the following error on the initial scan:
```open helmrepo/xxx.tgz: too many open files in system```

I haven't dug through the code yet, but my assumption is that when it first scans the repo it open each file and doesn't close it before opening the next, or it parallelizes the scan and opens so many files that it hits the file descriptor limit. For now, I have bumped by ulimit to get around this error, but I may not be able to do that in production.

**Comment by jdolitsky** [comment]
@sjthespian thank you for opening this.

You are correct that it parallelizes the scan. We may want to add a configurable limit for this. 

@davidovich any comments to add here?

**Comment by sjthespian** [comment]
@jdolitsky / @davidovich  It looks like it is even more of an issue if I run chartmuseum in a container. Even with ulimit set to increase nofiles, it fails with a thread limit trying to scan our repo. It looks like it definitely needs some command line options for parallelism.
```
2018-03-05T19:45:49.765Z        DEBUG   Fetching chart list from storage
2018-03-05T19:45:58.304Z        DEBUG   Regenerating index.yaml
2018-03-05T19:45:58.304Z        DEBUG   Loading charts packages from storage (this could take awhile)   {"total": 15945}
runtime: program exceeds 10000-thread limit
fatal error: thread exhaustion

runtime stack:
runtime.throw(0xf3bd33, 0x11)
        /usr/local/go/src/runtime/panic.go:596 +0x95
runtime.checkmcount()
        /usr/local/go/src/runtime/proc.go:519 +0xa4
runtime.mcommoninit(0xc43dfb5000)
        /usr/local/go/src/runtime/proc.go:539 +0xa0
runtime.allocm(0xc420014000, 0x0, 0xc43dfab000)
        /usr/local/go/src/runtime/proc.go:1341 +0x99
runtime.newm(0x0, 0xc420014000)
        /usr/local/go/src/runtime/proc.go:1611 +0x39
runtime.startm(0xc420014000, 0x14ce700)
        /usr/local/go/src/runtime/proc.go:1698 +0x17e
runtime.handoffp(0xc420014000)
        /usr/local/go/src/runtime/proc.go:1725 +0x55
runtime.retake(0x3ece48ade7b2, 0x2)
        /usr/local/go/src/runtime/proc.go:3909 +0x15d
runtime.sysmon()
        /usr/local/go/src/runtime/proc.go:3836 +0x1ea
runtime.mstart1()
        /usr/local/go/src/runtime/proc.go:1179 +0x11e
runtime.mstart()
        /usr/local/go/src/runtime/proc.go:1149 +0x64

```

**Comment by sjthespian** [comment]
If I'm looking at the correct code, I'm assuming you would have to add a counter to limit the parallel processes around https://github.com/kubernetes-helm/chartmuseum/blob/16021add17ffdc61824832435ba4a61d7fedab6a/pkg/chartmuseum/cache.go#L218 ?

**Comment by jdolitsky** [comment]
@sjthespian that is correct

**Comment by sjthespian** [comment]
I'm still a go novice, but would the following do the trick? (assuming indexLimit is a command line argument value with a reasonable default):
```
        limiter := make(chan bool, indexLimit)
	for _, object := range objects {
                limiter <- true
		go func(o storage.Object) {
                        defer func() { <-limiter }()
			select {
			case <-ctx.Done():
				return
			default:
				chartVersion, err := server.getObjectChartVersion(o, true)
				if err != nil {
					err = server.checkInvalidChartPackageError(log, o, err, "added")
					if err != nil {
						cancel()
					}
				}
				cvChan <- cvResult{chartVersion, err}
			}
		}(object)
	}
        // Wait for remaining func() calls to terminate
        for i := 0; i < cap(limiter); i++ {
            limiter <- true
        }
```

**Comment by jdolitsky** [comment]
@sjthespian ping me on k8s slack and lets work through this

**Comment by sjthespian** [comment]
I have a working piece of code, I'll fork the repo and send a PR once I figure out what to call the command line argument and get that working. Here's the diff using a hard-coded limit of 250:
```
iff --git a/pkg/chartmuseum/cache.go b/pkg/chartmuseum/cache.go
index a79b573..3ae8db6 100644
--- a/pkg/chartmuseum/cache.go
+++ b/pkg/chartmuseum/cache.go
@@ -214,8 +214,12 @@ func (server *Server) addIndexObjectsAsync(log loggingFn, index *repo.Index, obj
        ctx, cancel := context.WithCancel(context.Background())
        defer cancel()

+       indexLimit := 250
+       limiter := make(chan bool, indexLimit)
        for _, object := range objects {
+               limiter <- true
                go func(o storage.Object) {
+                       <-limiter
                        select {
                        case <-ctx.Done():
                                return
@@ -231,6 +235,10 @@ func (server *Server) addIndexObjectsAsync(log loggingFn, index *repo.Index, obj
                        }
                }(object)
        }
+       // Wait for remaining func() calls to terminate
+        for i := 0; i < cap(limiter); i++ {
+            limiter <- true
+        }

        for validCount := 0; validCount < numObjects; validCount++ {
                cvRes := <-cvChan
```

**Comment by jdolitsky** [comment]
Fixed by #60

**Comment by jdolitsky** [comment]
Fixed by #60
